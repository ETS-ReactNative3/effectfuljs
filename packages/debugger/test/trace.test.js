require("./setupTrace");
const trace = require("./run").bind(null, true);
const D = require("../main");

describe("time traveling", function() {
  test("Object tracing", function() {
    const s = Symbol("S");
    const v = D.wrap({ a: "A", b: "B", c: "C", [s]: "S" });
    const protoUnwrapped = {};
    const proto = D.wrap(protoUnwrapped);
    const orig = Object.setPrototypeOf({ ...v }, proto);
    Object.setPrototypeOf(v, proto);
    expect(Object.getPrototypeOf(v)).toBe(proto);
    let save = D.Trace.checkpoint();
    delete v.a;
    v.a = "a";
    delete v.b;
    v[s] = "s";
    Object.setPrototypeOf(v, null);
    expect(v).toStrictEqual(
      Object.assign(Object.create(null), {
        a: "a",
        c: "C",
        [s]: "s",
        [D.Trace.traceDataSymbol]: v[D.Trace.traceDataSymbol]
      })
    );
    D.Trace.undo(save);
    save = D.Trace.checkpoint();
    expect(v).toStrictEqual(orig);
  });
  test("Array tracing", function() {
    const arr = D.wrap([1, 2, 3, 4, 5]);
    const orig = [...arr];
    let save = D.Trace.checkpoint();
    arr.splice(2, 2, "A", "B", "C");
    expect([...arr]).toStrictEqual([1, 2, "A", "B", "C", 5]);
    D.Trace.undo(save);
    save = D.Trace.checkpoint();
    expect([...arr]).toStrictEqual(orig);
    arr.push("Z");
    arr.unshift("A");
    expect([...arr]).toStrictEqual(["A", 1, 2, 3, 4, 5, "Z"]);
    D.Trace.undo(save);
    save = D.Trace.checkpoint();
    expect([...arr]).toEqual(orig);
    arr[0] = "A";
    arr[4] = "Z";
    expect([...arr]).toEqual(["A", 2, 3, 4, "Z"]);
    D.Trace.undo(save);
    expect([...arr]).toEqual(orig);
  });
  test("Map tracing", function() {
    const map = D.wrap(new Map([[1, "A"], [2, "B"], [3, "C"]]));
    map.deb_hi = "there";
    const orig = [...map];
    let save = D.Trace.checkpoint();
    map.deb_hi = "THERE";
    map.set(1, "D").delete(2);
    map.delete(3);
    map.set(3, "E").set(4, "F");
    D.Trace.undo(save);
    expect(map.has(4)).toBe(false);
    expect(map.deb_hi).toBe("there");
    expect(map.get(3)).toBe("C");
    save = D.Trace.checkpoint();
    expect([...map]).toEqual(orig);
    map.clear();
    D.Trace.undo(save);
    save = D.Trace.checkpoint();
    expect([...map]).toEqual(orig);
    map.clear();
    map.set(2, "E");
    map.set(1, "F");
    D.Trace.undo(save);
    save = D.Trace.checkpoint();
    expect([...map]).toEqual(orig);
  });
  test("Set tracing", function() {
    const set = D.wrap(new Set([1, 2, 3]));
    const orig = [...set];
    set.deb_hi = "there";
    let save = D.Trace.checkpoint();
    set.deb_hi = "THERE";
    set.delete(1);
    set.delete(3);
    set.add(4);
    set.add(2);
    set.add(1);
    D.Trace.undo(save);
    save = D.Trace.checkpoint();
    expect(set.has(4)).toBe(false);
    expect(set.deb_hi).toBe("there");
    expect(set.has(3)).toBe(true);
    expect([...set]).toEqual(orig);
    set.clear();
    D.Trace.undo(save);
    save = D.Trace.checkpoint();
    expect([...set]).toEqual(orig);
    set.clear();
    set.add(8);
    D.Trace.undo(save);
    save = D.Trace.checkpoint();
    expect([...set]).toEqual(orig);
  });
  test("captured state", function() {
    global.console = { log: jest.fn() };
    D.evalThunk(require("./__fixtures__/counters"));
    const mod = trace();
    mod.getCounter();
    const [counter, tr, dir, set] = trace(true);
    console.log("INIT", tr, dir, set);
    let stepNum = 0;
    function counterStep() {
      counter();
      console.log(
        "STEP",
        stepNum++,
        trace(true),
        JSON.stringify({ tr, dir: [...dir], set: [...set] })
      );
    }
    counterStep();
    counterStep();
    counterStep();
    console.log("SAVE", tr, dir, set);
    const save = D.Trace.checkpoint();
    counterStep();
    counterStep();
    counterStep();
    counterStep();
    console.log("UNDO");
    D.Trace.undo(save);
    const save1 = D.Trace.checkpoint();
    counterStep();
    counterStep();
    console.log("SAVE2");
    const save2 = D.Trace.checkpoint();
    counterStep();
    counterStep();
    console.log("UNDO SAVE2");
    D.Trace.undo(save2);
    counterStep();
    counterStep();
    console.log("UNDO SAVE");
    try {
      D.Trace.undo(save);
    } catch (e) {
      console.log(e);
    }
    counter();
    console.log(trace(true));
    console.log("UNDO SAVE1");
    D.Trace.undo(save1);
    counter();
    console.log(trace());
    expect(console.log.mock.calls).toMatchSnapshot();
  });
});
