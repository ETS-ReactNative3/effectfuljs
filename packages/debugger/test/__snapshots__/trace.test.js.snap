// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`time traveling captured state 1`] = `
Array [
  Array [
    "INIT",
    Array [
      Object {
        "i": 0,
        "u": 4,
        Symbol(@effectful/debugger/traceData): TraceData {
          "lastUpdate": Object {
            "next": null,
            "prev": Object {
              "next": [Circular],
              "prev": Object {
                "next": [Circular],
                "prev": null,
                "undone": false,
                "version": 0,
              },
              "undone": false,
              "version": 1,
            },
            "undone": false,
            "version": 2,
          },
          "meta": Object {
            "proto": null,
            "wrap": [Function],
          },
          "prototypeChanged": false,
          "proxy": [Circular],
          "target": Object {
            "i": 0,
            "u": 4,
            Symbol(@effectful/debugger/traceData): [Circular],
          },
          "version": 2,
        },
      },
      Object {
        "i": 1,
        "u": 14,
        Symbol(@effectful/debugger/traceData): TraceData {
          "lastUpdate": Object {
            "next": null,
            "prev": Object {
              "next": [Circular],
              "prev": Object {
                "next": [Circular],
                "prev": null,
                "undone": false,
                "version": 0,
              },
              "undone": false,
              "version": 1,
            },
            "undone": false,
            "version": 2,
          },
          "meta": Object {
            "proto": null,
            "wrap": [Function],
          },
          "prototypeChanged": false,
          "proxy": [Circular],
          "target": Object {
            "i": 1,
            "u": 14,
            Symbol(@effectful/debugger/traceData): [Circular],
          },
          "version": 2,
        },
      },
      Object {
        "i": 2,
        "u": 24,
        Symbol(@effectful/debugger/traceData): TraceData {
          "lastUpdate": Object {
            "next": null,
            "prev": Object {
              "next": [Circular],
              "prev": Object {
                "next": [Circular],
                "prev": null,
                "undone": false,
                "version": 0,
              },
              "undone": false,
              "version": 1,
            },
            "undone": false,
            "version": 2,
          },
          "meta": Object {
            "proto": null,
            "wrap": [Function],
          },
          "prototypeChanged": false,
          "proxy": [Circular],
          "target": Object {
            "i": 2,
            "u": 24,
            Symbol(@effectful/debugger/traceData): [Circular],
          },
          "version": 2,
        },
      },
      Object {
        "i": 3,
        "u": 34,
        Symbol(@effectful/debugger/traceData): TraceData {
          "lastUpdate": Object {
            "next": null,
            "prev": Object {
              "next": [Circular],
              "prev": Object {
                "next": [Circular],
                "prev": null,
                "undone": false,
                "version": 0,
              },
              "undone": false,
              "version": 1,
            },
            "undone": false,
            "version": 2,
          },
          "meta": Object {
            "proto": null,
            "wrap": [Function],
          },
          "prototypeChanged": false,
          "proxy": [Circular],
          "target": Object {
            "i": 3,
            "u": 34,
            Symbol(@effectful/debugger/traceData): [Circular],
          },
          "version": 2,
        },
      },
      Object {
        "i": 4,
        "u": 44,
        Symbol(@effectful/debugger/traceData): TraceData {
          "lastUpdate": Object {
            "next": null,
            "prev": Object {
              "next": [Circular],
              "prev": Object {
                "next": [Circular],
                "prev": null,
                "undone": false,
                "version": 0,
              },
              "undone": false,
              "version": 1,
            },
            "undone": false,
            "version": 2,
          },
          "meta": Object {
            "proto": null,
            "wrap": [Function],
          },
          "prototypeChanged": false,
          "proxy": [Circular],
          "target": Object {
            "i": 4,
            "u": 44,
            Symbol(@effectful/debugger/traceData): [Circular],
          },
          "version": 2,
        },
      },
    ],
    Map {
      Object {
        "i": 0,
        "u": 4,
        Symbol(@effectful/debugger/traceData): TraceData {
          "lastUpdate": Object {
            "next": null,
            "prev": Object {
              "next": [Circular],
              "prev": Object {
                "next": [Circular],
                "prev": null,
                "undone": false,
                "version": 0,
              },
              "undone": false,
              "version": 1,
            },
            "undone": false,
            "version": 2,
          },
          "meta": Object {
            "proto": null,
            "wrap": [Function],
          },
          "prototypeChanged": false,
          "proxy": [Circular],
          "target": Object {
            "i": 0,
            "u": 4,
            Symbol(@effectful/debugger/traceData): [Circular],
          },
          "version": 2,
        },
      } => 0,
      Object {
        "i": 1,
        "u": 14,
        Symbol(@effectful/debugger/traceData): TraceData {
          "lastUpdate": Object {
            "next": null,
            "prev": Object {
              "next": [Circular],
              "prev": Object {
                "next": [Circular],
                "prev": null,
                "undone": false,
                "version": 0,
              },
              "undone": false,
              "version": 1,
            },
            "undone": false,
            "version": 2,
          },
          "meta": Object {
            "proto": null,
            "wrap": [Function],
          },
          "prototypeChanged": false,
          "proxy": [Circular],
          "target": Object {
            "i": 1,
            "u": 14,
            Symbol(@effectful/debugger/traceData): [Circular],
          },
          "version": 2,
        },
      } => 1,
      Object {
        "i": 2,
        "u": 24,
        Symbol(@effectful/debugger/traceData): TraceData {
          "lastUpdate": Object {
            "next": null,
            "prev": Object {
              "next": [Circular],
              "prev": Object {
                "next": [Circular],
                "prev": null,
                "undone": false,
                "version": 0,
              },
              "undone": false,
              "version": 1,
            },
            "undone": false,
            "version": 2,
          },
          "meta": Object {
            "proto": null,
            "wrap": [Function],
          },
          "prototypeChanged": false,
          "proxy": [Circular],
          "target": Object {
            "i": 2,
            "u": 24,
            Symbol(@effectful/debugger/traceData): [Circular],
          },
          "version": 2,
        },
      } => 2,
      Object {
        "i": 3,
        "u": 34,
        Symbol(@effectful/debugger/traceData): TraceData {
          "lastUpdate": Object {
            "next": null,
            "prev": Object {
              "next": [Circular],
              "prev": Object {
                "next": [Circular],
                "prev": null,
                "undone": false,
                "version": 0,
              },
              "undone": false,
              "version": 1,
            },
            "undone": false,
            "version": 2,
          },
          "meta": Object {
            "proto": null,
            "wrap": [Function],
          },
          "prototypeChanged": false,
          "proxy": [Circular],
          "target": Object {
            "i": 3,
            "u": 34,
            Symbol(@effectful/debugger/traceData): [Circular],
          },
          "version": 2,
        },
      } => 3,
      Object {
        "i": 4,
        "u": 44,
        Symbol(@effectful/debugger/traceData): TraceData {
          "lastUpdate": Object {
            "next": null,
            "prev": Object {
              "next": [Circular],
              "prev": Object {
                "next": [Circular],
                "prev": null,
                "undone": false,
                "version": 0,
              },
              "undone": false,
              "version": 1,
            },
            "undone": false,
            "version": 2,
          },
          "meta": Object {
            "proto": null,
            "wrap": [Function],
          },
          "prototypeChanged": false,
          "proxy": [Circular],
          "target": Object {
            "i": 4,
            "u": 44,
            Symbol(@effectful/debugger/traceData): [Circular],
          },
          "version": 2,
        },
      } => 4,
    },
    Set {
      0,
      1,
      2,
      3,
      4,
    },
  ],
  Array [
    "STEP",
    0,
    0,
    "{\\"tr\\":[{\\"i\\":0,\\"u\\":0}],\\"dir\\":[[{\\"i\\":0,\\"u\\":0},0]],\\"set\\":[0]}",
  ],
  Array [
    "STEP",
    1,
    1,
    "{\\"tr\\":[{\\"i\\":0,\\"u\\":1},{\\"i\\":1,\\"u\\":11}],\\"dir\\":[[{\\"i\\":0,\\"u\\":1},0],[{\\"i\\":1,\\"u\\":11},1]],\\"set\\":[0,1]}",
  ],
  Array [
    "STEP",
    2,
    2,
    "{\\"tr\\":[{\\"i\\":0,\\"u\\":2},{\\"i\\":1,\\"u\\":12},{\\"i\\":2,\\"u\\":22}],\\"dir\\":[[{\\"i\\":0,\\"u\\":2},0],[{\\"i\\":1,\\"u\\":12},1],[{\\"i\\":2,\\"u\\":22},2]],\\"set\\":[0,1,2]}",
  ],
  Array [
    "SAVE",
    Array [
      Object {
        "i": 0,
        "u": 4,
        Symbol(@effectful/debugger/traceData): TraceData {
          "lastUpdate": Object {
            "next": null,
            "prev": Object {
              "next": [Circular],
              "prev": Object {
                "next": [Circular],
                "prev": null,
                "undone": false,
                "version": 0,
              },
              "undone": false,
              "version": 1,
            },
            "undone": false,
            "version": 2,
          },
          "meta": Object {
            "proto": null,
            "wrap": [Function],
          },
          "prototypeChanged": false,
          "proxy": [Circular],
          "target": Object {
            "i": 0,
            "u": 4,
            Symbol(@effectful/debugger/traceData): [Circular],
          },
          "version": 2,
        },
      },
      Object {
        "i": 1,
        "u": 14,
        Symbol(@effectful/debugger/traceData): TraceData {
          "lastUpdate": Object {
            "next": null,
            "prev": Object {
              "next": [Circular],
              "prev": Object {
                "next": [Circular],
                "prev": null,
                "undone": false,
                "version": 0,
              },
              "undone": false,
              "version": 1,
            },
            "undone": false,
            "version": 2,
          },
          "meta": Object {
            "proto": null,
            "wrap": [Function],
          },
          "prototypeChanged": false,
          "proxy": [Circular],
          "target": Object {
            "i": 1,
            "u": 14,
            Symbol(@effectful/debugger/traceData): [Circular],
          },
          "version": 2,
        },
      },
      Object {
        "i": 2,
        "u": 24,
        Symbol(@effectful/debugger/traceData): TraceData {
          "lastUpdate": Object {
            "next": null,
            "prev": Object {
              "next": [Circular],
              "prev": Object {
                "next": [Circular],
                "prev": null,
                "undone": false,
                "version": 0,
              },
              "undone": false,
              "version": 1,
            },
            "undone": false,
            "version": 2,
          },
          "meta": Object {
            "proto": null,
            "wrap": [Function],
          },
          "prototypeChanged": false,
          "proxy": [Circular],
          "target": Object {
            "i": 2,
            "u": 24,
            Symbol(@effectful/debugger/traceData): [Circular],
          },
          "version": 2,
        },
      },
      Object {
        "i": 3,
        "u": 34,
        Symbol(@effectful/debugger/traceData): TraceData {
          "lastUpdate": Object {
            "next": null,
            "prev": Object {
              "next": [Circular],
              "prev": Object {
                "next": [Circular],
                "prev": null,
                "undone": false,
                "version": 0,
              },
              "undone": false,
              "version": 1,
            },
            "undone": false,
            "version": 2,
          },
          "meta": Object {
            "proto": null,
            "wrap": [Function],
          },
          "prototypeChanged": false,
          "proxy": [Circular],
          "target": Object {
            "i": 3,
            "u": 34,
            Symbol(@effectful/debugger/traceData): [Circular],
          },
          "version": 2,
        },
      },
      Object {
        "i": 4,
        "u": 44,
        Symbol(@effectful/debugger/traceData): TraceData {
          "lastUpdate": Object {
            "next": null,
            "prev": Object {
              "next": [Circular],
              "prev": Object {
                "next": [Circular],
                "prev": null,
                "undone": false,
                "version": 0,
              },
              "undone": false,
              "version": 1,
            },
            "undone": false,
            "version": 2,
          },
          "meta": Object {
            "proto": null,
            "wrap": [Function],
          },
          "prototypeChanged": false,
          "proxy": [Circular],
          "target": Object {
            "i": 4,
            "u": 44,
            Symbol(@effectful/debugger/traceData): [Circular],
          },
          "version": 2,
        },
      },
    ],
    Map {
      Object {
        "i": 0,
        "u": 4,
        Symbol(@effectful/debugger/traceData): TraceData {
          "lastUpdate": Object {
            "next": null,
            "prev": Object {
              "next": [Circular],
              "prev": Object {
                "next": [Circular],
                "prev": null,
                "undone": false,
                "version": 0,
              },
              "undone": false,
              "version": 1,
            },
            "undone": false,
            "version": 2,
          },
          "meta": Object {
            "proto": null,
            "wrap": [Function],
          },
          "prototypeChanged": false,
          "proxy": [Circular],
          "target": Object {
            "i": 0,
            "u": 4,
            Symbol(@effectful/debugger/traceData): [Circular],
          },
          "version": 2,
        },
      } => 0,
      Object {
        "i": 1,
        "u": 14,
        Symbol(@effectful/debugger/traceData): TraceData {
          "lastUpdate": Object {
            "next": null,
            "prev": Object {
              "next": [Circular],
              "prev": Object {
                "next": [Circular],
                "prev": null,
                "undone": false,
                "version": 0,
              },
              "undone": false,
              "version": 1,
            },
            "undone": false,
            "version": 2,
          },
          "meta": Object {
            "proto": null,
            "wrap": [Function],
          },
          "prototypeChanged": false,
          "proxy": [Circular],
          "target": Object {
            "i": 1,
            "u": 14,
            Symbol(@effectful/debugger/traceData): [Circular],
          },
          "version": 2,
        },
      } => 1,
      Object {
        "i": 2,
        "u": 24,
        Symbol(@effectful/debugger/traceData): TraceData {
          "lastUpdate": Object {
            "next": null,
            "prev": Object {
              "next": [Circular],
              "prev": Object {
                "next": [Circular],
                "prev": null,
                "undone": false,
                "version": 0,
              },
              "undone": false,
              "version": 1,
            },
            "undone": false,
            "version": 2,
          },
          "meta": Object {
            "proto": null,
            "wrap": [Function],
          },
          "prototypeChanged": false,
          "proxy": [Circular],
          "target": Object {
            "i": 2,
            "u": 24,
            Symbol(@effectful/debugger/traceData): [Circular],
          },
          "version": 2,
        },
      } => 2,
      Object {
        "i": 3,
        "u": 34,
        Symbol(@effectful/debugger/traceData): TraceData {
          "lastUpdate": Object {
            "next": null,
            "prev": Object {
              "next": [Circular],
              "prev": Object {
                "next": [Circular],
                "prev": null,
                "undone": false,
                "version": 0,
              },
              "undone": false,
              "version": 1,
            },
            "undone": false,
            "version": 2,
          },
          "meta": Object {
            "proto": null,
            "wrap": [Function],
          },
          "prototypeChanged": false,
          "proxy": [Circular],
          "target": Object {
            "i": 3,
            "u": 34,
            Symbol(@effectful/debugger/traceData): [Circular],
          },
          "version": 2,
        },
      } => 3,
      Object {
        "i": 4,
        "u": 44,
        Symbol(@effectful/debugger/traceData): TraceData {
          "lastUpdate": Object {
            "next": null,
            "prev": Object {
              "next": [Circular],
              "prev": Object {
                "next": [Circular],
                "prev": null,
                "undone": false,
                "version": 0,
              },
              "undone": false,
              "version": 1,
            },
            "undone": false,
            "version": 2,
          },
          "meta": Object {
            "proto": null,
            "wrap": [Function],
          },
          "prototypeChanged": false,
          "proxy": [Circular],
          "target": Object {
            "i": 4,
            "u": 44,
            Symbol(@effectful/debugger/traceData): [Circular],
          },
          "version": 2,
        },
      } => 4,
    },
    Set {
      0,
      1,
      2,
      3,
      4,
    },
  ],
  Array [
    "STEP",
    3,
    3,
    "{\\"tr\\":[{\\"i\\":0,\\"u\\":3},{\\"i\\":1,\\"u\\":13},{\\"i\\":2,\\"u\\":23},{\\"i\\":3,\\"u\\":33}],\\"dir\\":[[{\\"i\\":0,\\"u\\":3},0],[{\\"i\\":1,\\"u\\":13},1],[{\\"i\\":2,\\"u\\":23},2],[{\\"i\\":3,\\"u\\":33},3]],\\"set\\":[0,1,2,3]}",
  ],
  Array [
    "STEP",
    4,
    4,
    "{\\"tr\\":[{\\"i\\":0,\\"u\\":4},{\\"i\\":1,\\"u\\":14},{\\"i\\":2,\\"u\\":24},{\\"i\\":3,\\"u\\":34},{\\"i\\":4,\\"u\\":44}],\\"dir\\":[[{\\"i\\":0,\\"u\\":4},0],[{\\"i\\":1,\\"u\\":14},1],[{\\"i\\":2,\\"u\\":24},2],[{\\"i\\":3,\\"u\\":34},3],[{\\"i\\":4,\\"u\\":44},4]],\\"set\\":[0,1,2,3,4]}",
  ],
  Array [
    "STEP",
    5,
    5,
    "{\\"tr\\":[{\\"i\\":0,\\"u\\":5},{\\"i\\":1,\\"u\\":15},{\\"i\\":2,\\"u\\":25},{\\"i\\":3,\\"u\\":35},{\\"i\\":4,\\"u\\":45},{\\"i\\":5,\\"u\\":55}],\\"dir\\":[[{\\"i\\":0,\\"u\\":5},0],[{\\"i\\":1,\\"u\\":15},1],[{\\"i\\":2,\\"u\\":25},2],[{\\"i\\":3,\\"u\\":35},3],[{\\"i\\":4,\\"u\\":45},4],[{\\"i\\":5,\\"u\\":55},5]],\\"set\\":[0,1,2,3,4,5]}",
  ],
  Array [
    "STEP",
    6,
    6,
    "{\\"tr\\":[{\\"i\\":0,\\"u\\":6},{\\"i\\":1,\\"u\\":16},{\\"i\\":2,\\"u\\":26},{\\"i\\":3,\\"u\\":36},{\\"i\\":4,\\"u\\":46},{\\"i\\":5,\\"u\\":56},{\\"i\\":6,\\"u\\":66}],\\"dir\\":[[{\\"i\\":0,\\"u\\":6},0],[{\\"i\\":1,\\"u\\":16},1],[{\\"i\\":2,\\"u\\":26},2],[{\\"i\\":3,\\"u\\":36},3],[{\\"i\\":4,\\"u\\":46},4],[{\\"i\\":5,\\"u\\":56},5],[{\\"i\\":6,\\"u\\":66},6]],\\"set\\":[0,1,2,3,4,5,6]}",
  ],
  Array [
    "UNDO",
  ],
  Array [
    "STEP",
    7,
    3,
    "{\\"tr\\":[{\\"i\\":0,\\"u\\":3},{\\"i\\":1,\\"u\\":13},{\\"i\\":2,\\"u\\":23},{\\"i\\":3,\\"u\\":33}],\\"dir\\":[[{\\"i\\":0,\\"u\\":3},0],[{\\"i\\":1,\\"u\\":13},1],[{\\"i\\":2,\\"u\\":23},2],[{\\"i\\":3,\\"u\\":33},3]],\\"set\\":[0,1,2,3]}",
  ],
  Array [
    "STEP",
    8,
    4,
    "{\\"tr\\":[{\\"i\\":0,\\"u\\":4},{\\"i\\":1,\\"u\\":14},{\\"i\\":2,\\"u\\":24},{\\"i\\":3,\\"u\\":34},{\\"i\\":4,\\"u\\":44}],\\"dir\\":[[{\\"i\\":0,\\"u\\":4},0],[{\\"i\\":1,\\"u\\":14},1],[{\\"i\\":2,\\"u\\":24},2],[{\\"i\\":3,\\"u\\":34},3],[{\\"i\\":4,\\"u\\":44},4]],\\"set\\":[0,1,2,3,4]}",
  ],
  Array [
    "SAVE2",
  ],
  Array [
    "STEP",
    9,
    5,
    "{\\"tr\\":[{\\"i\\":0,\\"u\\":5},{\\"i\\":1,\\"u\\":15},{\\"i\\":2,\\"u\\":25},{\\"i\\":3,\\"u\\":35},{\\"i\\":4,\\"u\\":45},{\\"i\\":5,\\"u\\":55}],\\"dir\\":[[{\\"i\\":0,\\"u\\":5},0],[{\\"i\\":1,\\"u\\":15},1],[{\\"i\\":2,\\"u\\":25},2],[{\\"i\\":3,\\"u\\":35},3],[{\\"i\\":4,\\"u\\":45},4],[{\\"i\\":5,\\"u\\":55},5]],\\"set\\":[0,1,2,3,4,5]}",
  ],
  Array [
    "STEP",
    10,
    6,
    "{\\"tr\\":[{\\"i\\":0,\\"u\\":6},{\\"i\\":1,\\"u\\":16},{\\"i\\":2,\\"u\\":26},{\\"i\\":3,\\"u\\":36},{\\"i\\":4,\\"u\\":46},{\\"i\\":5,\\"u\\":56},{\\"i\\":6,\\"u\\":66}],\\"dir\\":[[{\\"i\\":0,\\"u\\":6},0],[{\\"i\\":1,\\"u\\":16},1],[{\\"i\\":2,\\"u\\":26},2],[{\\"i\\":3,\\"u\\":36},3],[{\\"i\\":4,\\"u\\":46},4],[{\\"i\\":5,\\"u\\":56},5],[{\\"i\\":6,\\"u\\":66},6]],\\"set\\":[0,1,2,3,4,5,6]}",
  ],
  Array [
    "UNDO SAVE2",
  ],
  Array [
    "STEP",
    11,
    5,
    "{\\"tr\\":[{\\"i\\":0,\\"u\\":5},{\\"i\\":1,\\"u\\":15},{\\"i\\":2,\\"u\\":25},{\\"i\\":3,\\"u\\":35},{\\"i\\":4,\\"u\\":45},{\\"i\\":5,\\"u\\":55}],\\"dir\\":[[{\\"i\\":0,\\"u\\":5},0],[{\\"i\\":1,\\"u\\":15},1],[{\\"i\\":2,\\"u\\":25},2],[{\\"i\\":3,\\"u\\":35},3],[{\\"i\\":4,\\"u\\":45},4],[{\\"i\\":5,\\"u\\":55},5]],\\"set\\":[0,1,2,3,4,5]}",
  ],
  Array [
    "STEP",
    12,
    6,
    "{\\"tr\\":[{\\"i\\":0,\\"u\\":6},{\\"i\\":1,\\"u\\":16},{\\"i\\":2,\\"u\\":26},{\\"i\\":3,\\"u\\":36},{\\"i\\":4,\\"u\\":46},{\\"i\\":5,\\"u\\":56},{\\"i\\":6,\\"u\\":66}],\\"dir\\":[[{\\"i\\":0,\\"u\\":6},0],[{\\"i\\":1,\\"u\\":16},1],[{\\"i\\":2,\\"u\\":26},2],[{\\"i\\":3,\\"u\\":36},3],[{\\"i\\":4,\\"u\\":46},4],[{\\"i\\":5,\\"u\\":56},5],[{\\"i\\":6,\\"u\\":66},6]],\\"set\\":[0,1,2,3,4,5,6]}",
  ],
  Array [
    "UNDO SAVE",
  ],
  Array [
    3,
  ],
  Array [
    "UNDO SAVE1",
  ],
  Array [
    4,
  ],
]
`;
